stackslib/src/chainstate/stacks/boot/pox-4.clar

```diff
@@ -670,6 +680,44 @@
 
       (ok true)))
 
+;; Generate a message hash for validating a signer key.
+;; The message hash follows SIP018 for signing structured data. The structured data
+;; is the tuple `{ pox-addr: { version, hashbytes }, reward-cycle }`. The domain is
+;; `{ name: "pox-4-signer", version: "1.0.0", chain-id: chain-id }`.
+(define-read-only (get-signer-key-message-hash (pox-addr { version: (buff 1), hashbytes: (buff 32) }))
+  (let
+    (
+      (domain { name: "pox-4-signer", version: "1.0.0", chain-id: chain-id })
+      (data-hash (sha256 (unwrap-panic
+        (to-consensus-buff? { pox-addr: pox-addr, reward-cycle: (current-pox-reward-cycle) }))))
+      (domain-hash (sha256 (unwrap-panic (to-consensus-buff? domain))))
+    )
+    (sha256 (concat
+      SIP018_MSG_PREFIX
+      (concat domain-hash
+      data-hash)))
+  )
+)
+
+;; Verify a signature from the signing key for this specific stacker.
+;; See `get-signer-key-message-hash` for details on the message hash.
+;;
+;; Note that `reward-cycle` corresponds to the _current_ reward cycle,
+;; not the reward cycle at which the delegation will start.
+;; The public key is recovered from the signature and compared to `signer-key`.
+(define-read-only (verify-signer-key-sig (pox-addr { version: (buff 1), hashbytes: (buff 32) })
```

Comment:
Similarly, can this take the reward cycle as an argument, and pass it into `get-signer-key-message-hash`?

Reply:
Also, given the way this is used in `stack-aggregation-commit-indexed`, you'd actually need a caller-supplied reward cycle, since it'll be different than the current reward cycle.